hash/hash_comparator.go

```diff
@@ -1,55 +1,169 @@
 	"golang.org/x/crypto/bcrypt"
+	"golang.org/x/crypto/sha3"
 
 	"github.com/ory/kratos/driver/config"
 )
 
-var ErrUnknownHashAlgorithm = errors.New("unknown hash algorithm")
+var ErrUnknownHashAlgorithm = fmt.Errorf("unknown hash algorithm")
+var ErrUnknownHashFormat = fmt.Errorf("unknown hash format")
+var ErrEmptyHashCompare = fmt.Errorf("empty hash provided")
+var ErrEmptyPasswordCompare = fmt.Errorf("empty password provided")
 
-func Compare(ctx context.Context, password []byte, hash []byte) error {
-	if IsBcryptHash(hash) {
-		return CompareBcrypt(ctx, password, hash)
-	} else if IsArgon2idHash(hash) {
-		return CompareArgon2id(ctx, password, hash)
-	} else {
+var hashSeparator = []byte("$")
+var bcryptLegacyPrefix = regexp.MustCompile(`^2[abzy]?$`)
+
+func Compare(ctx context.Context, cfg *config.Config, password []byte, hash []byte) error {
+	algorithm, realHash, err := ParsePasswordHash(hash)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	switch {
+	case bytes.Equal(algorithm, Argon2AlgorithmId):
+		return CompareArgon2id(ctx, cfg, password, realHash)
+	case bytes.Equal(algorithm, BcryptAlgorithmId):
+		return CompareBcrypt(ctx, cfg, password, realHash)
+	case bytes.Equal(algorithm, BcryptAESAlgorithmId):
+		return CompareBcryptAes(ctx, cfg, password, realHash)
+	default:
 		return ErrUnknownHashAlgorithm
 	}
 }
 
-func CompareBcrypt(_ context.Context, password []byte, hash []byte) error {
+func CompareBcrypt(_ context.Context, _ *config.Config, password []byte, hash []byte) error {
 	if err := validateBcryptPasswordLength(password); err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	err := bcrypt.CompareHashAndPassword(hash, password)
 	if err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	return nil
 }
 
-func CompareArgon2id(_ context.Context, password []byte, hash []byte) error {
+func IsPasswordHash(hash []byte) bool {
+	_, _, err := ParsePasswordHash(hash)
+	return err == nil
+}
+
+func ParsePasswordHash(input []byte) (algorithm, hash []byte, err error) {
+	hashParts := bytes.SplitN(input, hashSeparator, 3)
+	if len(hashParts) != 3 {
+		err = errors.WithStack(ErrUnknownHashFormat)
+		return
+	}
+
+	hash = append(hashSeparator, hashParts[2]...)
+	switch {
+	case bytes.Equal(hashParts[1], Argon2AlgorithmId):
+		algorithm = Argon2AlgorithmId
+		return
+	case bcryptLegacyPrefix.Match(hashParts[1]):
+		hash = input
+		fallthrough
+	case bytes.Equal(hashParts[1], BcryptAlgorithmId):
+		algorithm = BcryptAlgorithmId
+		return
+	case bytes.Equal(hashParts[1], BcryptAESAlgorithmId):
+		algorithm = BcryptAESAlgorithmId
+		return
+	default:
+		err = ErrUnknownHashAlgorithm
+		return
+	}
+}
+
+// aes256Decrypt decrypts data using 256-bit AES-GCM.  This both hides the content of
+// the data and provides a check that it hasn't been altered. Expects input
+// form nonce|ciphertext|tag where '|' indicates concatenation.
+func aes256Decrypt(ciphertext []byte, key *[32]byte) (plaintext []byte, err error) {
```

Comment:
I would move this to the aes hasher file and also mirror a comment I made earlier suggesting to use cryptopasta (as an import) instead of copy/pasting or writing our own here. I’ve implemented AES GCM before (with stdlib) and it‘s very easy to make small mistakes in the encoding or key preparations.

Reply:
Done in b9aa9132.